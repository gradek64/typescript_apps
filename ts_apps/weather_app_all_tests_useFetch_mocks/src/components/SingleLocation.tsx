import React from 'react';
import { useFetch } from '../services/useFetch'


//custom 
import './singleLocation.css'
import { getDateFormat, dateToTimeDashFormat } from '../misc/formatDate'

const SingleLocation = () => {

  const location = { london: { lon: '-0.118092', lat: '51.509865' } }
  const { data, loading, error } = useFetch({ url: `https://weatherbit-v1-mashape.p.rapidapi.com/current?lon=${location.london.lon}&lat=${location.london.lat}` })

  const formatDate = data && data.reduce((acc, { datetime }) => { acc = dateToTimeDashFormat(datetime); return acc }, '')

  return (
    <div className='singleLocation'>
      {loading && 'loading'}
      {error && error}
      {data && data.map(({ city_name, temp, clouds, weather, country_code }, i: number) =>
        <div key={i} className='singleUnit'>
          <div className='cityInfo'>
            <h2>{city_name},{country_code}</h2>
            <p>
              {`${formatDate && getDateFormat(formatDate, 'day')} 
              ${formatDate && getDateFormat(formatDate, 'dateNumber')}
              ${formatDate && getDateFormat(formatDate, 'month')}`}
            </p>
          </div>
          <div className='weatherInfo'>
            <span className="icon-alone" aria-hidden="true" data-icon="&#x231a;"></span>
            <p>{temp}<span className='metric'>Â°C</span></p>
            <span>{weather.description}</span>
          </div>

        </div>
      )}
    </div>
  )
}
export default SingleLocation