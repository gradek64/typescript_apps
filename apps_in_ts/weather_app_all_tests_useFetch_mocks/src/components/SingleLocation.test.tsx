import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import SingleLocation from './SingleLocation';

const mockData = {
  data: [{
    city_name: 'London',
    datetime: '2021-02-10:12',
    temp: 3,
    weather: {
      description: 'cloudy'
    },
  }]
}

const mockJsonPromise = Promise.resolve(mockData);
const mockFetchPromise = Promise.resolve({
  json: () => mockJsonPromise,
});

describe('@render', () => {
  it('Component is loading', () => {
    render(<SingleLocation />);
    const linkElement = screen.getByText(/loading/i);
    expect(linkElement).toBeInTheDocument();

  });
  it('Component has loaded data', async () => {
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    const { asFragment } = render(<SingleLocation />);
    //after http response
    const comp = await waitFor(() => asFragment())
    expect(comp).toMatchSnapshot();
  });
})

