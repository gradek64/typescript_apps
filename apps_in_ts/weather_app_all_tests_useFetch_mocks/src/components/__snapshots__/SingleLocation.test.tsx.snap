// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@render Component has loaded data 1`] = `
<DocumentFragment>
  <div
    class="singleLocation"
  >
    loading
  </div>
</DocumentFragment>
`;
